src/file_path.rs

```diff
@@ -0,0 +1,160 @@
+use std::fmt::{self, Debug, Formatter, Write};
+use std::mem::transmute;
+
+#[cfg(any(windows, unix, target_os = "wasi"))]
+use std::{ffi::OsStr, path::Path};
+
+#[cfg(unix)]
+use std::os::unix::ffi::OsStrExt;
+#[cfg(target_os = "wasi")]
+use std::os::wasi::ffi::OsStrExt;
+
+use crate::Error;
+
+/// Path to a file, required to be UTF-8 on windows
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct FilePath(#[cfg(windows)] str, #[cfg(not(windows))] [u8]);
+
+impl FilePath {
+    pub fn new<P: AsRef<FilePath>>(p: &P) -> &Self {
+        p.as_ref()
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.as_ref()
+    }
+}
+
+impl Debug for FilePath {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_char('"')?;
+
+        #[cfg(windows)]
+        write!(f, "{}", self.0.escape_debug())?;
+
+        #[cfg(not(windows))]
+        for chunk in self.0.utf8_chunks() {
+            write!(f, "{}", chunk.valid().escape_debug())?;
+            write!(f, "{}", chunk.invalid().escape_ascii())?;
+        }
+
+        f.write_char('"')
+    }
+}
+
+impl AsRef<FilePath> for str {
+    fn as_ref(&self) -> &FilePath {
+        #[cfg(windows)]
+        // SAFETY: On windows FilePath is a str. As `self` is a str as well
+        // and FilePath is repr(transparent) this is sound.
+        unsafe {
+            transmute(self)
+        }
```

Comment:
I assume we're not able to do `&FilePath(*self)` 'cause that would be returning a local value?